Finish project part

Omar622 opened the pull request:
### description
- create project model
- make crud endpoints
- make endpoint tests
- have not integrated yet with authorization middleware

### Issues
- [project part](https://github.com/codescalersinternships/Flyspray/issues/1)
- [add a wrapper for handlers](https://github.com/codescalersinternships/Flyspray/issues/22)

Omar622 committed:
Finish project part

- create project model
- make crud endpoints
- make endpoint tests

rawdaGastan commented on backend/app/project_handler.go:
```diff
@@ -0,0 +1,67 @@
+	Name string `json:"name" binding:"required"`
+}
+
+func createProject(ctx *gin.Context, db models.DBClient) {
```
handlers should be on app struct. so you can use the DB without passing it

rawdaGastan commented on backend/app/project_handler.go:
```diff
@@ -0,0 +1,67 @@
+func createProject(ctx *gin.Context, db models.DBClient) {
+	var input createProjectInput
+
+	if err := ctx.BindJSON(&input); err != nil {
```
we should log errors

rawdaGastan commented on backend/app/project_handler.go:
```diff
@@ -0,0 +1,67 @@
+
+	if result := db.Client.Create(&newProject); result.Error != nil {
+		log.Fatal(result.Error)
+		ctx.Status(http.StatusInternalServerError)
```
we should return error msgs for the user

rawdaGastan commented on backend/app/app.go:
```diff
@@ -33,6 +33,18 @@ func (a *App) Run(port int) error {
 	a.router = gin.Default()
 
 	// set routes here
+	a.router.POST("project", func(ctx *gin.Context) {
```
what about having a new function to register handlers?

rawdaGastan replied:
group project handlers together.

rawdaGastan commented on backend/app/project_handler.go:
```diff
@@ -0,0 +1,67 @@
+		return
+	}
+
+	ctx.IndentedJSON(http.StatusCreated, newProject)
```
what about having a response and error response structs for your handlers?

rawdaGastan commented on backend/app/project_handler.go:
```diff
@@ -0,0 +1,67 @@
+func getProject(ctx *gin.Context, db models.DBClient) {
+	id := ctx.Param("id")
+
+	project := models.Project{}
```
should have DB functions for crud operations
check all

rawdaGastan commented on backend/app/project_handler.go:
```diff
@@ -0,0 +1,67 @@
+	result := db.Client.First(&project, id)
+
+	if result.Error == gorm.ErrRecordNotFound {
+		ctx.Status(http.StatusNotFound)
```
return msgs
check all

rawdaGastan commented on backend/app/project_handler.go:
```diff
@@ -0,0 +1,67 @@
+}
+
+func getProjects(ctx *gin.Context, db models.DBClient) {
+	userId := ctx.Query("userid")
```
you should have more filters, `name` for example

rawdaGastan commented on backend/app/project_handler.go:
```diff
@@ -0,0 +1,67 @@
+}
+
+func deleteProject(ctx *gin.Context, db models.DBClient) {
+	// TODO: get user id from authorization middleware and check if user has access to delete the project
```
I dont think we will need user ID

rawdaGastan commented on backend/app/app.go:
```diff
@@ -33,6 +33,18 @@ func (a *App) Run(port int) error {
+	})
+	a.router.DELETE("project/:id", func(ctx *gin.Context) {
+		deleteProject(ctx, a.client)
+	})
```
add update handler

rawdaGastan commented on backend/models/project.go:
```diff
@@ -0,0 +1,10 @@
+
+import "time"
+
+type Project struct {
```
add DB functions for crud operation here and use it directly in the handlers

rawdaGastan commented on backend/models/project.go:
```diff
@@ -0,0 +1,10 @@
+import "time"
+
+type Project struct {
+	ID        uint
```
add it as a primary key, also add json tags

Omar622 committed:
Add update project endpoint & refactor handlers

refactoring:
- move db logic to separate functions
- make struct type for response(err, ok)
- add tests for update project endpoint
- group routes

Omar622 committed:
Make db operation on DBClient struct

AbdelrahmanElawady commented on backend/app/project_handler.go:
```diff
@@ -0,0 +1,126 @@
+	var input createProjectInput
+
+	if err := ctx.BindJSON(&input); err != nil {
+		log.Println("project name must be specified")
```
you should log `err` itself and add context to what was happening during error. there are several reasons it could fail other than `name not specified`

AbdelrahmanElawady commented on backend/app/project_handler.go:
```diff
@@ -0,0 +1,126 @@
+	if err := ctx.BindJSON(&input); err != nil {
+		log.Println("project name must be specified")
+		ctx.IndentedJSON(http.StatusBadRequest,
+			responseErr{Message: "name must be specified"},
```
same here

AbdelrahmanElawady commented on backend/app/project_handler.go:
```diff
@@ -0,0 +1,126 @@
+		)
+		return
+	}
+
```
what if `err` is not `record not found`?

AbdelrahmanElawady commented on backend/app/project_handler.go:
```diff
@@ -0,0 +1,126 @@
+}
+
+func (a *App) getProject(ctx *gin.Context) {
+	// TODO: add middleware to check if user is signed in
```
does the user need to sign in to see a project?

Omar622 replied:
yes, rawda told me this

AbdelrahmanElawady commented on backend/app/project_handler.go:
```diff
@@ -0,0 +1,126 @@
+		)
+		return
+	}
+
```
same here, check for other errors.

AbdelrahmanElawady commented on backend/app/project_handler_test.go:
```diff
@@ -0,0 +1,185 @@
+		app.router.ServeHTTP(w, req)
+
+		assert.Equal(t, http.StatusBadRequest, w.Code)
+	})
```
add test for an input that is not a `createProjectInput`

AbdelrahmanElawady commented on backend/app/project_handler.go:
```diff
@@ -0,0 +1,126 @@
+func (a *App) updateProject(ctx *gin.Context) {
+	// TODO: get user id from authorization middleware and check if user has access to update the project
+	id := ctx.Param("id")
+	var input createProjectInput
```
why is `createProjectInput` is the input to update project?

AbdelrahmanElawady commented on backend/app/project_handler_test.go:
```diff
@@ -0,0 +1,185 @@
+		payload, err := json.Marshal(input)
+		assert.Nil(t, err)
+
+		req, err := http.NewRequest("PUT", "/project/1", bytes.NewBuffer(payload))
```
tests should not depend on each other this way, each tests should be able to run on its own. state should be isolated between tests and tests should not leave artifacts behind them.

AbdelrahmanElawady commented on backend/models/project.go:
```diff
@@ -0,0 +1,84 @@
+import (
+	"time"
+)
+
```
docstring

AbdelrahmanElawady commented on backend/models/project.go:
```diff
@@ -0,0 +1,84 @@
+
+type Project struct {
+	ID        uint      `json:"id" gorm:"primary_key; unique"`
+	Name      string    `json:"name" validate:"nonzero"`
```
can two projects have the same name?

AbdelrahmanElawady commented on backend/models/project.go:
```diff
@@ -0,0 +1,84 @@
+// UpdateProject updates project
+func (d *DBClient) UpdateProject(id string, updatedProject Project) (Project, error) {
+	project := Project{}
+	result := d.Client.First(&project, id)
+	if result.Error != nil {
+		return Project{}, result.Error
+	}
+
+	project.Name = updatedProject.Name
+	d.Client.Save(&project)
```
`gorm` can update without fetching the resource first

AbdelrahmanElawady commented on backend/models/project.go:
```diff
@@ -0,0 +1,84 @@
+// DeleteProject deletes project by id
+func (d *DBClient) DeleteProject(id string) {
+	project := Project{}
+	d.Client.Delete(&project, id)
```
this can return an error, why is it not handled?

AbdelrahmanElawady commented on backend/models/project.go:
```diff
@@ -0,0 +1,84 @@
+		return projects
+	}
+
+	d.Client.Where("owner_id = ? and name = ? and created_at > ?", userId, projectName, date).Find(&projects)
```
check error here

AbdelrahmanElawady commented on backend/models/project.go:
```diff
@@ -0,0 +1,84 @@
+func (d *DBClient) FilterProjects(userId, projectName, date string) []Project {
+	projects := []Project{}
+
+	if userId == "" && projectName == "" && date == "" {
+		d.Client.Find(&projects)
+		return projects
+	}
+
+	if userId == "" && projectName == "" {
+		d.Client.Where("created_at > ?", date).Find(&projects)
+		return projects
+	}
+	if userId == "" && date == "" {
+		d.Client.Where("name = ?", projectName).Find(&projects)
+		return projects
+	}
+	if projectName == "" && date == "" {
+		d.Client.Where("owner_id = ?", userId).Find(&projects)
+		return projects
+	}
+
+	if userId == "" {
+		d.Client.Where("name = ? and created_at > ?", projectName, date).Find(&projects)
+		return projects
+	}
+	if projectName == "" {
+		d.Client.Where("owner_id = ? and created_at > ?", userId, date).Find(&projects)
+		return projects
+	}
+	if date == "" {
+		d.Client.Where("owner_id = ? and name = ?", userId, projectName).Find(&projects)
```
can you build the condition based on inputs instead of having separate db query and return

AbdelrahmanElawady replied:
also check if `gorm` supports filtering directly

AbdelrahmanElawady reviewed (changes requested):
nice work, some minor comments :)

Omar622 committed:
Add handlers' wrapper & refactor tests

- make tests independent and remove its artifacts

Omar622 committed:
Add conflict response to the wrapper

AbdelrahmanElawady commented on backend/app/project_handler.go:
```diff
@@ -0,0 +1,130 @@
+	}
+
+	// check if project name exists before
+	if _, err := a.client.GetProjectByName(input.Name); err != gorm.ErrRecordNotFound {
+		log.Error().Err(err).Send()
+		return nil, BadRequest(errors.New("project name must be unique"))
```
if it is found should it be bad request?

AbdelrahmanElawady commented on backend/app/project_handler.go:
```diff
@@ -0,0 +1,130 @@
+	}
+
+	// check if project name exists before
+	if _, err := a.client.GetProjectByName(input.Name); err != gorm.ErrRecordNotFound {
+		log.Error().Err(err).Send()
+		return nil, BadRequest(errors.New("project name must be unique"))
```
same here, also the error could be nil or it could be another error (like can't connect to db) so not sure this is the best way to handle it

AbdelrahmanElawady replied:
you can just update the project without getting it first and check affected rows to determine if it is not found

AbdelrahmanElawady commented on backend/app/project_handler.go:
```diff
@@ -0,0 +1,130 @@
+	}
+
+	return ResponseMsg{
+		Message: "projects is retrieved successfully",
```
`are`

AbdelrahmanElawady commented on backend/app/project_handler_test.go:
```diff
@@ -0,0 +1,286 @@
+)
+
+func TestCreateProject(t *testing.T) {
+	dbFilePath := "./test.db"
```
make sure to remove the db after testing, maybe in a temp dir

AbdelrahmanElawady commented on backend/app/wrapper.go:
```diff
@@ -0,0 +1,153 @@
+}
+
+// Error generic error response
+func Error(err error, code ...int) Response {
```
why is code a slice, will it ever be more than one?

AbdelrahmanElawady commented on backend/app/wrapper.go:
```diff
@@ -0,0 +1,153 @@
+
+// InternalServerError result
+func InternalServerError(err error) Response {
+	return Error(err, 0)
```
why not 500?

Omar622 committed:
Refactor wrapper

- update unnecessary slice in `Error` func
- return 500 status code instead of 0 in `InternalServerError` func

Omar622 committed:
Handle not found when delete project

- add test for it

Omar622 committed:
Handle not found in update project

- implement repeated project name in a better way
  in update, create project endpoints
- add tests

Omar622 committed:
Create temp db when testing & make table tests

Omar622 committed:
Handle repeated project name in a better way

rawdaGastan commented on backend/app/project_handler.go:
```diff
@@ -0,0 +1,152 @@
+		return nil, BadRequest(errors.New("failed to read project data"))
+	}
+
+	// TODO: get user id from authorization middleware and assign it to OwnerId
```
change hardcoded user IDs

rawdaGastan commented on backend/app/project_handler.go:
```diff
@@ -0,0 +1,152 @@
+	}
+	if err != gorm.ErrRecordNotFound { // there is some error other than not found
+		log.Error().Err(err).Send()
+		return nil, InternalServerError(errors.New("failed to create project"))
```
I think we should return `internal server error` as a msg for all InternalServerErrors

rawdaGastan commented on backend/app/project_handler.go:
```diff
@@ -0,0 +1,152 @@
+	}
+	if err != nil && err != gorm.ErrRecordNotFound { // there is some error and it is other than not found
+		log.Error().Err(err).Send()
+		return nil, InternalServerError(errors.New("failed to create project"))
```
the msg is incorrect. InternalServerError msg will be better

rawdaGastan commented on backend/models/db.go:
```diff
@@ -22,7 +22,10 @@ type DBClient struct {
 // Migrate makes migrations for the database
-func (d *DBClient) Migrate() error{
+func (d *DBClient) Migrate() error {
+	if err := d.Client.AutoMigrate(&Project{}); err != nil {
```
return d.Client.AutoMigrate(&Project{}) directly

rawdaGastan commented on backend/models/project.go:
```diff
@@ -0,0 +1,71 @@
+type Project struct {
+	ID        uint      `json:"id" gorm:"primary_key; unique"`
+	Name      string    `json:"name" validate:"nonzero" gorm:"unique"`
+	OwnerId   uint      `json:"owner_id" validate:"nonzero"`
```
change to `OwnerID` 
OwnerID is string

AbdelrahmanElawady commented on backend/app/project_handler_test.go:
```diff
@@ -0,0 +1,263 @@
+package app
```
table based tests may need some changes as you are not checking if the db contains the data you sent or updated or not, consider altering the structure or not using table based tests

Omar622 committed:
Refactor err message

- change internal server error message
- make OwnerID string

Omar622 committed:
Test changes in db & response body

Omar622 committed:
Replace hardcoded userID with middleware's userID

AbdelrahmanElawady commented on backend/app/project_handler.go:
```diff
@@ -63,16 +67,38 @@ func (a *App) updateProject(ctx *gin.Context) (interface{}, Response) {
+	if err == nil && fmt.Sprint(p.ID) != id {      // another project is found has the same name as the updated name
 		return nil, BadRequest(errors.New("project name must be unique"))
 	}
 	if err != nil && err != gorm.ErrRecordNotFound { // there is some error and it is other than not found
```
no need for `err != nil`

AbdelrahmanElawady replied:
misread second condition, you are correct

AbdelrahmanElawady reviewed (approved):
Great work!

rawdaGastan commented on backend/app/project_handler.go:
```diff
@@ -0,0 +1,197 @@
+		return nil, InternalServerError(errors.New("internal server error"))
+	}
+
+	newProject, err = a.client.CreateProject(newProject)
```
I think we can remove the `GetProjectByName` part and check if the error is duplicate key `err == gorm.ErrDuplicatedKey`

apply fot update function also

rawdaGastan commented on backend/app/project_handler.go:
```diff
@@ -0,0 +1,197 @@
+
+	if err != nil {
+		log.Error().Err(err).Send()
+		return nil, InternalServerError(errors.New("internal server error"))
```
what about creating an error msg for `internal server error` to be used in all functions?

Omar622 committed:
Make var for intervalServerErr msg
